### PR TITLE
Move record delimiting into ColumnReader (#4365)

### LINKED ISSUE #4365: List-of-String Array panics in the presence of row filters
**Describe the bug**

Reading Parquet files that have a List array with items of type `Utf8`, panic in the presence of a filter on another column due to a length mismatch of arrays when building the underlying StructArray.

**To Reproduce**

This is a pretty small example I built:
```rust
use arrow::array::{ListBuilder, StringBuilder, UInt32Array};
use arrow::datatypes::{DataType, Field, Schema};
use arrow::record_batch::RecordBatch;
use futures::StreamExt;
use std::sync::Arc;

use parquet::arrow::arrow_reader::{ArrowPredicateFn, RowFilter};
use parquet::arrow::async_reader::AsyncFileReader;
use parquet::arrow::ParquetRecordBatchStreamBuilder;
use parquet::arrow::{ArrowWriter, ProjectionMask};
use parquet::file::properties::WriterProperties;
use std::error::Error;

use tempfile::NamedTempFile;

const BATCH_SIZE: usize = 1024;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    let schema = Arc::new(Schema::new(vec![
        Field::new(
            "list",
            DataType::List(Arc::new(Field::new("item", DataType::Utf8, true))),
            false,
        ),
        Field::new("numbers", DataType::UInt32, false),
    ]));
    let temp_file = NamedTempFile::new()?;

    let mut writer = ArrowWriter::try_new(
        temp_file.reopen()?,
        schema.clone(),
        Some(WriterProperties::builder().build()),
    )?;
    for _ in 0..2 {
        let mut list_a_builder = ListBuilder::new(StringBuilder::new());
        for i in 0..1024 {
            list_a_builder.values().append_value(format!("{i}"));

            list_a_builder.append(true);
        }
        let batch = RecordBatch::try_new(
            schema.clone(),
            vec![
                Arc::new(list_a_builder.finish()),
                Arc::new(UInt32Array::from_iter_values(
                    (0..BATCH_SIZE).map(|n| n as u32),
                )),
            ],
        )?;
        writer.write(&batch)?;
    }
    let _metadata = writer.close()?;

    let mut file = tokio::fs::File::open(temp_file.path()).await.unwrap();

    let parquet_metadata = file.get_metadata().await.unwrap();
    let file_metadata = parquet_metadata.file_metadata();
    let schema_descriptor = file_metadata.schema_descr();

    // We filter on the numerical column
    let row_filter = RowFilter::new(vec![Box::new(ArrowPredicateFn::new(
        ProjectionMask::leaves(schema_descriptor, vec![schema.fields().len() - 1]),
        |batch| arrow::compute::gt_dyn_scalar(batch.column(0), 100),
    ))]);

    // This is the key section - whether we materialize the list-of-strings column alone or with another column that was filtered
    #[cfg(feature = "with_bug")]
    let projection_mask = ProjectionMask::roots(schema_descriptor, [0, 1]); // Both columns

    #[cfg(not(feature = "with_bug"))]
    let projection_mask = ProjectionMask::roots(schema_descriptor, [0]); // Just the list of strings columns

    let mut reader = ParquetRecordBatchStreamBuilder::new(file)
        .await
        .unwrap()
        .with_row_filter(row_filter)
        .with_projection(projection_mask)
        .build()
        .unwrap();

    while let Some(rb) = reader.next().await {
        let rb = rb.unwrap();
        println!("count = {}", rb.num_rows())
    }

    Ok(())
}
```

the `Cargo.toml` file here looks like:
```toml
[package]
name = "reproduce"
version = "0.1.0"
edition = "2021"

[features]
with_bug = []

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
arrow = { version = "40", features = ["simd"] }
parquet = { version = "40", features = ["async"] }
tempfile = "*"
tokio = { version = "1.16.1", features = [
    "macros",
    "rt-multi-thread",
    "time",
    "fs",
] }
futures = "0.3.18"
```

run it with `cargo run --features with_bug`

**Expected behavior**

My understanding is that it should work, and the example even contains a case where it works.

**Additional context**

Would be glad to take on the effort, I'm just not exactly sure what's the right way to tackle this issue. Seems like it's somewhere relativly deep the the `Parquet`

### PR DESCRIPTION
**Pull request by tustvold**
_Draft as needs a bit more polish_

# Which issue does this PR close?

Closes #4365

# Rationale for this change
 

Previously when reading repeated fields, RecordReader would read in batches of at least 1024 levels. From this it would then split off the levels comprising the desired number of rows. Aside from being wasteful, and complex, this method causes issues for selection pushdown as it would need to skip into data that has already potentially been decoded.

# What changes are included in this PR?

This changes ColumnReader to delimit records, whilst this does potentially impose overheads on users using this API directly, I struggle to contrive a use-case where reading partial records would be desirable / correct. I therefore hypothesize such users are either not using repeated data (likely) or are already performing some sort of record buffering similar to what is now performed as a first-party API. 

# Are there any user-facing changes?

Technically no, however, marking this as breaking as it does alter the behaviour of ColumnReader, especially with regards to ill-formed level data as one might encounter in a test.

### CODE REVIEW

File: parquet/src/record/triplet.rs [hunk 1]
```diff
@@ -295,16 +295,16 @@ impl<T: DataType> TypedTripletIter<T> {
     fn read_next(&mut self) -> Result<bool> {
         self.curr_triplet_index += 1;
 
-        if self.curr_triplet_index >= self.triplets_left {
-            let (values_read, levels_read) = {
+        while self.curr_triplet_index >= self.triplets_left {
```
Comment (tustvold) [inline parquet/src/record/triplet.rs:298]: I hope to make it so that this change isn't necessary

File: parquet/src/column/reader/decoder.rs [hunk 2]
```diff
@@ -270,182 +297,227 @@ impl<T: DataType> ColumnValueDecoder for ColumnValueDecoderImpl<T> {
...
         }
 
         Ok((value_skip, level_skip))
     }
 }
 
-impl RepetitionLevelDecoder for ColumnLevelDecoderImpl {
-    fn skip_rep_levels(
+/// An implementation of [`RepetitionLevelDecoder`] for `[i16]`
+pub struct RepetitionLevelDecoderImpl {
```
Comment (tustvold) [inline parquet/src/column/reader/decoder.rs:389]: This is the main crux of the PR, move the record delimiting logic into the repetition level decoding process, instead of wrapped around ColumnReader

File: parquet/src/arrow/record_reader/mod.rs [hunk 3]
```diff
@@ -313,90 +231,29 @@ where
...
-    }
-
-    /// Inspects the buffered repetition levels in the range `self.num_values..self.values_written`
-    /// and returns the number of "complete" records along with the corresponding number of values
-    ///
-    /// If `end_of_column` is true it indicates that there are no further values for this
-    /// column chunk beyond what is currently in the buffers
-    ///
-    /// A "complete" record is one where the buffer contains a subsequent repetition level of 0
-    fn count_records(
```
Comment (tustvold) [inline parquet/src/arrow/record_reader/mod.rs:336]: This logic is now moved into RepetitionLevelDecoderImpl

File: parquet/src/arrow/arrow_reader/mod.rs [hunk 4]
```diff
@@ -2658,4 +2653,111 @@ mod tests {
         let total_rows: usize = reader.map(|r| r.unwrap().num_rows()).sum();
         assert_eq!(total_rows, 924 * 2);
     }
+
+    #[test]
+    fn test_list_selection_fuzz() {
```
Comment (tustvold) [inline parquet/src/arrow/arrow_reader/mod.rs:2658]: I'm quite pleased with this, the parquet reader is pretty sophisticated now :smile:

File: parquet/src/arrow/arrow_reader/mod.rs [hunk 5]
```diff
@@ -2610,4 +2611,153 @@ mod tests {
...
+                RowSelector::skip(100),
+                RowSelector::select(924),
+                RowSelector::skip(100),
+                RowSelector::select(924),
+            ]))
+            .build()
+            .unwrap();
+
+        let total_rows: usize = reader.map(|r| r.unwrap().num_rows()).sum();
+        assert_eq!(total_rows, 924 * 2);
```
Comment (alamb) [inline parquet/src/arrow/arrow_reader/mod.rs:2654]: I wonder if it is worth asserting the values come back as expected (you could make the expected results in the `for _ in 0..2` loop above)?

File: parquet/src/arrow/arrow_reader/mod.rs [hunk 6]
```diff
@@ -2610,4 +2611,153 @@ mod tests {
...
+                RowSelector::skip(100),
+                RowSelector::select(924),
+            ],
+            vec![
+                RowSelector::select(924),
+                RowSelector::skip(100),
+                RowSelector::select(924),
+                RowSelector::skip(100),
+            ],
+            vec![
```
Comment (alamb) [inline parquet/src/arrow/arrow_reader/mod.rs:2719]: I like this boundary condition testing

File: parquet/src/arrow/arrow_reader/mod.rs [hunk 7]
```diff
@@ -2610,4 +2611,153 @@ mod tests {
...
+                RowSelector::skip(1023),
+            ],
+        ];
+
+        for selection in cases {
+            let selection = RowSelection::from(selection);
+            let mut reader = ParquetRecordBatchReaderBuilder::try_new(buf.clone())
+                .unwrap()
+                .with_row_selection(selection.clone())
+                .with_batch_size(2048)
```
Comment (alamb) [inline parquet/src/arrow/arrow_reader/mod.rs:2738]: Given all the inputs are only 2048 rows in total (and thus in a single batch) would tests with different batch sizes that span the input (e.g `1000` for example) be a good idea?

File: parquet/src/arrow/record_reader/mod.rs [hunk 8]
```diff
@@ -201,31 +149,10 @@ where
...
-        self.num_records += buffered_records;
-        self.num_values += buffered_values;
-
-        let remaining = num_records - buffered_records;
-
-        if remaining == 0 {
-            return Ok(buffered_records);
+        match self.column_reader.as_mut() {
+            Some(reader) => reader.skip_records(num_records),
+            None => Ok(0),
```
Comment (alamb) [inline parquet/src/arrow/record_reader/mod.rs:154]: Is it intentional that this function no longer updates `self.num_records` and `self.num_values`

```
        self.num_records += buffered_records;
        self.num_values += buffered_values;
```

File: parquet/src/column/reader.rs [hunk 9]
```diff
@@ -195,99 +198,126 @@ where
...
+
+                    let read = reader.read_def_levels(
+                        out,
+                        total_levels_read..total_levels_read + levels_read,
+                    )?;
 
-                    if num_def_levels != iter_batch_size {
-                        return Err(general_err!("insufficient definition levels read from column - expected {}, got {}", iter_batch_size, num_def_levels));
+                    if read != levels_read {
+                        return Err(general_err!("insufficient definition levels read from column - expected {rep_levels}, got {read}"));
```
Comment (alamb) [inline parquet/src/column/reader.rs:291]: this happens with an invalid input, right?

File: parquet/src/column/reader/decoder.rs [hunk 10]
```diff
@@ -94,6 +105,22 @@ pub trait RepetitionLevelDecoder: ColumnLevelDecoder {
...
+    /// Read definition level data into `out[range]` returning the number of levels read
+    ///
+    /// `range` is provided by the caller to allow for types such as default-initialized `[T]`
+    /// that only track capacity and not length
+    ///
+    /// # Panics
+    ///
+    /// Implementations may panic if `range` overlaps with already written data
+    ///
+    // TODO: Should this return the number of nulls
```
Comment (alamb) [inline parquet/src/column/reader/decoder.rs:117]: is this still a TODO?

File: parquet/src/column/reader/decoder.rs [hunk 11]
```diff
@@ -270,182 +297,227 @@ impl<T: DataType> ColumnValueDecoder for ColumnValueDecoderImpl<T> {
...
-            break;
+            total_levels_read += levels_read;
+            total_records_read += records_read;
+            self.buffer_offset += levels_read;
+            self.has_partial = partial;
         }
+        Ok((total_records_read, total_levels_read))
+    }
+
+    fn skip_rep_levels(
```
Comment (alamb) [inline parquet/src/column/reader/decoder.rs:494]: it might help to note here in comments that it tries to read up to `num_records` and `num_levels` (whichever is greater) and  returns the (total_records_read, total_levels_read)

**Review by alamb — APPROVED** [review]
👏 

I can't say I really understand all of the record level decoding logic, but I did read it all, and I reviewed the tests carefully. Very well done 👍

File: parquet/src/arrow/record_reader/mod.rs (hunk 8 above)
Comment (tustvold) [reply parquet/src/arrow/record_reader/mod.rs:154]: Yes, RecordReader no longer performs this intermediate buffering

File: parquet/src/column/reader/decoder.rs (hunk 11 above)
Comment (tustvold) [reply parquet/src/column/reader/decoder.rs:494]: I think the docs on the trait implementation are sufficient

File: parquet/src/column/reader/decoder.rs (hunk 10 above)
Comment (tustvold) [reply parquet/src/column/reader/decoder.rs:117]: Yes

**Comment by zeevm** [comment]
@tustvold the new read_records() api returns the tuple (total_records_read, total_values_read, total_levels_read)

What's the purpose of total_levels_read? this will be either 0 is max def levels is 0, otherwise it should be the same as total_records_read, shouldn't it?

**Comment by tustvold** [comment]
For a repeated field the number of levels read may exceed the number of records read, as a record is delimited by a repetition level of 0

**Comment by zeevm** [comment]
but the '0' rep level is also associated with the first value of the next record, so the total_records should also include that shouldn't it?
Or will the read_records() impl will only read 'full records', not parts of records?

**Comment by alamb** [comment]
Follow on PR: https://github.com/apache/arrow-rs/pull/4540
